pkg/kube/controllers/statefulset/statefulset_rollout_reconciler.go

```diff
@@ -0,0 +1,225 @@
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"code.cloudfoundry.org/quarks-utils/pkg/pointers"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	"code.cloudfoundry.org/cf-operator/pkg/kube/apis"
+
+	"code.cloudfoundry.org/quarks-utils/pkg/config"
+	"code.cloudfoundry.org/quarks-utils/pkg/ctxlog"
+	"k8s.io/api/apps/v1beta2"
+	"k8s.io/apimachinery/pkg/runtime"
+	crc "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	rolloutStatePending       = "Pending"
+	rolloutStateCanary        = "Canary"
+	rolloutStateRollout       = "Rollout"
+	rolloutStateDone          = "Done"
+	rolloutStateFailed        = "Failed"
+	rolloutStateCanaryUpscale = "CanaryUpscale"
+)
+
+var (
+	// AnnotationCanaryRolloutEnabled if set to "true" canary behaviour is desired
+	AnnotationCanaryRolloutEnabled = fmt.Sprintf("%s/canary-rollout-enabled", apis.GroupName)
+	// AnnotationCanaryRollout is the state of the canary rollout of the stateful set
+	AnnotationCanaryRollout = fmt.Sprintf("%s/canary-rollout", apis.GroupName)
+	// AnnotationCanaryWatchTime is the max time for the canary update
+	AnnotationCanaryWatchTime = fmt.Sprintf("%s/canary-watch-time-ms", apis.GroupName)
+	// AnnotationUpdateWatchTime is the max time for the complete update
+	AnnotationUpdateWatchTime = fmt.Sprintf("%s/update-watch-time-ms", apis.GroupName)
+	// AnnotationUpdateStartTime is the timestamp when the update started
+	AnnotationUpdateStartTime = fmt.Sprintf("%s/update-start-time", apis.GroupName)
+)
+
+// NewStatefulSetRolloutReconciler returns a new reconcile.Reconciler
+func NewStatefulSetRolloutReconciler(ctx context.Context, config *config.Config, mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileStatefulSetRollout{
+		ctx:    ctx,
+		config: config,
+		client: mgr.GetClient(),
+		scheme: mgr.GetScheme(),
+	}
+}
+
+// ReconcileStatefulSetRollout reconciles an ExtendedStatefulSet object when references changes
+type ReconcileStatefulSetRollout struct {
+	ctx    context.Context
+	client crc.Client
+	scheme *runtime.Scheme
+	config *config.Config
+}
+
+// Reconcile cleans up old versions and volumeManagement statefulSet of the ExtendedStatefulSet
+func (r *ReconcileStatefulSetRollout) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+
+	// Set the ctx to be Background, as the top-level context for incoming requests.
+	ctx, cancel := context.WithTimeout(r.ctx, r.config.CtxTimeOut)
+	defer cancel()
+
+	ctxlog.Debug(ctx, "Reconciling StatefulSet ", request.NamespacedName)
+
+	statefulSet := v1beta2.StatefulSet{}
+
+	err := r.client.Get(ctx, request.NamespacedName, &statefulSet)
+	if err != nil {
+		ctxlog.Debug(ctx, "StatefulSet not found ", request.NamespacedName)
+		return reconcile.Result{}, err
+	}
+	var result reconcile.Result
+	var status = statefulSet.Annotations[AnnotationCanaryRollout]
```

Comment:
status shouldn't live in annotations, it should live in status object of the CRD
https://github.com/cloudfoundry-incubator/cf-operator/blob/master/pkg/kube/apis/quarksstatefulset/v1alpha1/types.go#L65

Reply:
@viovanov the controller is a _decorator_ for StatefulSets, so we should have the status on the StatefulSet. Reusing the StatefulSet status is dangerous though, hence the decision to go for annotation.

Reply:
Having it be independent from the qsts crd is not a requirement, but we can see how this goes.
I don't like not having the state in the qsts, but we can read it from the annotation.